File: 2025/day9/main.c
#include <stdio.h>
#include <stdlib.h>

int allowed(int points[1000][2], int n, int y, int x){
	int valid[] = {0,0,0,0};
	for (int i = 0; i <n; i++){
		int py = points[i][0];
		int px = points[i][1];

		if (py>=y && px>=x)
			valid[0] = 1;
		if (py>=y && px<=x)
			valid[1] = 1;
		if (py<=y && px>=x)
			valid[2] = 1;
		if (py<=y && px<=x)
			valid[3] = 1;
		if (valid[0]+valid[1]+valid[2]+valid[3] == 4){
			return 1;
		}
	}
	return 0;
}



int main(void) {
    FILE *fp = fopen("input2.txt", "r");

    int a, b;
    char line[256];

    int n = 496;
    int points[1000][2];
    for(int i = 0; i<n ;i++){
    	points[i][0]=0;
    	points[i][1]=0;
    }

    int i = 0;

    while (fgets(line, sizeof(line), fp)) {
        sscanf(line, "%d,%d", &a, &b);
        points[i][0] = a;
        points[i][1] = b;
        printf("%d,%d\n",a,b);
        i++;
    }
  
    fclose(fp);

    int max = 0;
    for (int i1 = 0; i1<n; i1++){
    	int y1 = points[i1][0];
    	int x1 = points[i1][1];

    	for (int i2 = 0; i2<i1; i2++){
    		int y2 = points[i2][0];
    		int x2 = points[i2][1];
    		//printf("comparing (%d, %d) - (%d, %d)\n",y1,x1,y2,x2);
    		int valid = 1;

    		int yii = y1<y2?y1:y2;
    		int xii = x1<x2?x1:x2;
    		int yia = y1>y2?y1:y2;
    		int xia = x1>x2?x1:x2;
    		int yi = yii;
    		while (valid && yi<yia){
	    		int xi = xii;
	    		while (valid && xi<xia){
	    			if (allowed(points, n, yi, xi) == 0)
    					valid = 0;
    				xi++;
    			}
    			yi ++;
    		}
    		if (valid==1){
    			int d = (abs(y2-y1)+1)*(abs(x2-x1)+1);
    			if (d > max){
    				max = d;
    				printf("%d\n",max);
    			}
    		}

    	}

    }


    return 0;
}

File: 2025/day9/main.py

filenames = ["input1.txt", "input2.txt"]


# part 1
for filename in filenames[2:]:

	m = [ [int(x) for x in line.strip().split(",")] for line in open(filename)]

	pwd = []
	for p1 in m:
		pwd.append( [ (abs(p1[0]-p2[0])+1) * (abs(p1[1]-p2[1])+1)  for p2 in m] )

	ma = 0
	for line in pwd:
		ma =max(ma, max(line))

	print(ma)



# part 2
def allowed(points, x,y):
	valid = [0,0,0,0]
	for py, px in points:
		if py>=y and px>=x:
			valid[0] = 1
		if py>=y and px<=x:
			valid[1] = 1
		if py<=y and px>=x:
			valid[2] = 1
		if py<=y and px<=x:
			valid[3] = 1
		if sum(valid) == 4:
			return True
	return False

for filename in filenames[1:]:

	m = [ [int(x) for x in line.strip().split(",")] for line in open(filename)]

	#tmp = [[0]*13 for _ in range(13)]
	#for y in range(13):
	#	for x in range(13):
	#		tmp[y][x] = "#" if allowed(m, y, x) else " "
	#for line in tmp:
	#	print(line)


	pwd = []
	for p1 in range(len(m)):
		print(p1/len(m))
		p1 = m[p1]
		tmp = [] 
		for p2 in m:
			valid = True
			yii, yia = min(p1[1], p2[1]), max(p1[1], p2[1])
			xii, xia = min(p1[0], p2[0]), max(p1[0], p2[0])
			yi = yii
			while valid and yi < yia:
				valid &= allowed(m, yi, xii)
				valid &= allowed(m, yi, xia)
				yi +=1
			xi = xii
			while valid and xi < xia:
				valid &= allowed(m, yii, xi)
				valid &= allowed(m, yia, xi)
				xi +=1
			if valid:
				tmp.append( (abs(p1[0]-p2[0])+1) * (abs(p1[1]-p2[1])+1))
			else:
				tmp.append(0)
		pwd.append(tmp)

	ma = 0
	for line in pwd:
		ma =max(ma, max(line))

	print(ma)

# 1506688 too low

# 4638696212 too high
# 2116295688 too high